### PR TITLE
sign binaries

### PR DESCRIPTION
**Pull request by jskelin**

### CODE REVIEW

File: .ci/releasePipeline.groovy [hunk 1]
```diff
@@ -141,3 +59,105 @@ pipeline {
...
+                                          secretValues: [
+                                              [envVar: 'username', vaultKey: 'username', isRequired: true],
+                                              [envVar: 'password', vaultKey: 'password', isRequired: true]]]]
+                ) {
+                    sh '''java -jar sign-client.jar \
+                        --action SIGN \
+                        --digestAlg SHA256 \
+                        --downloadPath ./build/signed \
+                        --kernelMode false  \
+                        --triggeringProject datasource-go \
```
Comment (UnseenWizzard) [inline .ci/releasePipeline.groovy:104]:
```suggestion
                        --triggeringProject monaco \
```

File: .ci/releasePipeline.groovy [hunk 2]
```diff
@@ -141,3 +65,122 @@ pipeline {
...
+                        --username $username \
+                        --password $password \
+                        $binarySource'''
+        }
+    }
+}
+
+
+
+def releaseBinaryToArtifactory(Map args = [binary: '', version: '', ctx: null]) {
```
Comment (UnseenWizzard) [inline .ci/releasePipeline.groovy:134]: please keep the non-specific 'ArtifactStorage' name

File: .ci/releasePipeline.groovy [hunk 3]
```diff
@@ -141,3 +65,161 @@ pipeline {
...
+                        --triggeringProject monaco \
+                        --triggeringVersion $version \
+                        --username $username \
+                        --password $password \
+                        $binarySource'''
+        }
+    }
+}
+
+void signLinuxBinaries(Map args) {
```
Comment (UnseenWizzard) [inline .ci/releasePipeline.groovy:133]: nit:
```suggestion
void signUnixBinaries(Map args) {
```

File: .ci/releasePipeline.groovy [hunk 4]
```diff
@@ -141,3 +65,161 @@ pipeline {
...
+    def shaSumFile = "SHA256SUMS"
```
Comment (UnseenWizzard) [inline .ci/releasePipeline.groovy:134]: as we create a file per binary I would suggest to call it by the binaries name
```suggestion
    def shaSumFile = "${args.binary}.sha"
```

A SHASUMS file like e.g. used by Ubuntu makes more sense if the file contains the sums for several files

**Review by UnseenWizzard — CHANGES_REQUESTED** [review]
please add the container build back in
other comments are mostly nitpicks/small improvement suggestions :)

File: .ci/releasePipeline.groovy (hunk 2 above)
Comment (jskelin) [reply .ci/releasePipeline.groovy]: done

**Review by jskelin — COMMENTED** [review]
create and push docker container added

File: .ci/releasePipeline.groovy [hunk 5]
```diff
@@ -141,3 +127,175 @@ pipeline {
...
+
+void signWinBinaries(Map args = [source: null, version: null, destDir: null, projectName: null]) {
+    stage('Sign binaries') {
+        signToSignService(source: args.source, version: args.version, destDir: args.destDir, signAction: "SIGN", projectName: args.projectName)
+    }
+}
+
+void createShaSum(Map args = [source: null, dest: null]) {
+    stage('Compute SHA sum') {
+        sh "sha256sum ${args.source} | sed 's, .*/, ,' > ${args.dest}"
```
Comment (UnseenWizzard) [inline .ci/releasePipeline.groovy:171]: what exactly does that sed expression replace? I assume some line ending characters that caused issues with sign service when I tested - and why not just have shasum omit the line break (-z) as my version had

Did you test the verification steps locally?

File: .ci/releasePipeline.groovy [hunk 6]
```diff
@@ -141,3 +127,175 @@ pipeline {
...
+                         --user "$username":"$password"
+                         --upload-file $source
+                         --fail-with-body
+                   '''.replaceAll("\n", " ").replaceAll(/ +/, " ")
+            }
+        }
+    }
+}
+
+void signToSignService(Map args = [source: null, version: null, destDir: null, signAction: null, projectName: null]) {
```
Comment (UnseenWizzard) [inline .ci/releasePipeline.groovy:208]: nit: 

```suggestion
void signWithSignService(Map args = [source: null, version: null, destDir: null, signAction: null, projectName: null]) {
```

reads more clearly/correct to me

**Review by UnseenWizzard — CHANGES_REQUESTED** [review]
If I didn't miss something, I think the github release steps will fail (release ID not parsed)

File: .ci/releasePipeline.groovy [hunk 7]
```diff
@@ -141,3 +127,175 @@ pipeline {
...
+                    } finally {
+                        sh 'docker logout $registry'
+                    }
+                }
+            }
+        }
+    }
+}
+
+def getReleaseIdFormGithub(Map args = [version: null]) {
```
[the remark this answers is not in the document]
Comment (jskelin) [reply .ci/releasePipeline.groovy:253]: done

File: .ci/releasePipeline.groovy [hunk 8]
```diff
@@ -38,92 +16,100 @@ pipeline {
...
-                                    sh 'docker push $REGISTRY_PATH/monaco/dynatrace-monitoring-as-code:$VERSION'
+                        excludes {
+                            exclude {
+                                axis {
+                                    name 'OS'
+                                    values 'windows', 'linux'
+                                }
+                                axis {
+                                    name 'ARCH'
+                                    values 'arm64'
```
[the remark this answers is not in the document]
Comment (jskelin) [reply .ci/releasePipeline.groovy:55]: added linux-arm64 to build

File: .ci/releasePipeline.groovy (hunk 5 above)
Comment (jskelin) [reply .ci/releasePipeline.groovy:171]: in a path of file for which are sha sum computed, this removes leading part of path if `sha256sum` command isn't executed in same directory

It is checked once, but will do it again when token will be ready, and the whole procedure will pass from the beginning to the end

File: .ci/releasePipeline.groovy (hunk 6 above)
Comment (jskelin) [reply .ci/releasePipeline.groovy:208]: done

**Review by UnseenWizzard — APPROVED** [review]
Thanks for all the improvements! :tada:

File: .ci/releasePipeline.groovy [hunk 9]
```diff
@@ -141,3 +127,175 @@ pipeline {
...
+                             --fail-with-body
+                             --data \'{
+                                        "tag_name":"v$version",
+                                        "target_commitish":"main",
+                                        "name":"$version",
+                                        "draft":true,
+                                        "prerelease":false,
+                                        "generate_release_notes":true
+                                    }\'
+                        '''.replaceAll("\n", " ").replaceAll(/ +/, " "))
```
[the remark this answers is not in the document]
Comment (jskelin) [reply .ci/releasePipeline.groovy:275]: it is done programmatically

File: .ci/releasePipeline.groovy [hunk 10]
```diff
@@ -38,92 +16,100 @@ pipeline {
...
+                                                pushToDynatraceStorage(source: release, name: release, version: version)
+                                                pushToGithub(source: release, name: release, releaseID: releaseId)
+                                                break
+                                            case "linux":
+                                            case "darwin":
+                                                def release = "${PROJECT}-${env.OS}-${env.ARCH}" //name of the tar archive
+
+                                                sh "rm -rf ${release}" //ensure to delete old build
+
+                                                buildBinary(command: release, version: version, dest: "${release}/${PROJECT}")
```
[the remark this answers is not in the document]
Comment (jskelin) [reply .ci/releasePipeline.groovy:103]: exactly

here is how it looks in repo.
and content of tar file (e.g. monaco-linux-386.tar) is:
```
tar -tvf monaco-linux-386.tar 
-rwxr-xr-x 0/0         8420616 2023-03-07 14:00 monaco
-rw-r--r-- 0/0            4643 2023-03-07 14:00 monaco-linux-386.sha
```

if name of *.sha needs to be changed, it is easy change

**Comment by UnseenWizzard** [comment]
closing in favor of uploading seperate files: https://github.com/Dynatrace/dynatrace-configuration-as-code/pull/872

**Pull request closed without merging**